snapshot panic - test run contains new node delegation tests

[https://jenkins.ops.vega.xyz/job/common/job/system-tests-nightly/95/artifact/system-tes[…]dator/vega-validator-3.stderr-2022-09-14T15%3A23%3A49Z.log](https://jenkins.ops.vega.xyz/job/common/job/system-tests-nightly/95/artifact/system-tests-full-s/testnet/logs/testnet-nodeset-validators-3-validator/vega-validator-3.stderr-2022-09-14T15%3A23%3A49Z.log)

panic

```
panic: failed to save snapshot to disk

goroutine 147497 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc0044f75c0, {0xc0039a2bc0, 0x1, 0x1})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.21.0/zapcore/entry.go:232 +0x44c
go.uber.org/zap.(*Logger).Panic(0xc000553680?, {0x249adbf?, 0xc75f3d8bf0f4c623?}, {0xc0039a2bc0, 0x1, 0x1})
	/jenkins/GOPATH/pkg/mod/go.uber.org/zap@v1.21.0/logger.go:230 +0x59
code.vegaprotocol.io/vega/core/snapshot.(*Engine).snapshotNow.func2()
	/jenkins/workspace/common/system-tests-wrapper/vega/core/snapshot/engine.go:855 +0x14b
created by code.vegaprotocol.io/vega/core/snapshot.(*Engine).snapshotNow
	/jenkins/workspace/common/system-tests-wrapper/vega/core/snapshot/engine.go:865 +0xde5
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@
 - [6153](https://github.com/vegaprotocol/vega/issues/6153) - Display UI friendly logs when calling `session.send_transaction`
 
 ### 🐛 Fixes
+- [6183](https://github.com/vegaprotocol/vega/issues/6183) - Shutdown blockchain before protocol services
 - [6148](https://github.com/vegaprotocol/vega/issues/6148) - Fix API descriptions for typos
 - [6156](https://github.com/vegaprotocol/vega/issues/6156) - Return only delegations for the specific node in `graphql` node delegation query
 - [6175](https://github.com/vegaprotocol/vega/issues/6175) - Fix `datanode` updating node public key on key rotation

--- a/cmd/vega/node/node.go
+++ b/cmd/vega/node/node.go
@@ -205,14 +205,14 @@ func (n *Command) startProtocolUpgrade(version string) {
 }
 
 func (n *Command) Stop() error {
+	if n.blockchainServer != nil {
+		n.blockchainServer.Stop()
+	}
 	if n.protocol != nil {
 		n.protocol.Stop()
 	}
 	if n.grpcServer != nil {
 		n.grpcServer.Stop()
-	}
-	if n.blockchainServer != nil {
-		n.blockchainServer.Stop()
 	}
 	if n.proxyServer != nil {
 		n.proxyServer.Stop()
@@ -316,7 +316,7 @@ func (n *Command) startBlockchain(tmHome, network, networkURL string) error {
 		if err != nil {
 			return err
 		}
-		n.blockchainServer = blockchain.NewServer(n.tmNode)
+		n.blockchainServer = blockchain.NewServer(n.Log, n.tmNode)
 		// initialise the client
 		client, err := n.tmNode.GetClient()
 		if err != nil {
@@ -328,7 +328,7 @@ func (n *Command) startBlockchain(tmHome, network, networkURL string) error {
 		// nullchain acts as both the client and the server because its does everything
 		n.nullBlockchain = nullchain.NewClient(n.Log, n.conf.Blockchain.Null)
 		n.nullBlockchain.SetABCIApp(n.abciApp)
-		n.blockchainServer = blockchain.NewServer(n.nullBlockchain)
+		n.blockchainServer = blockchain.NewServer(n.Log, n.nullBlockchain)
 		// n.blockchainClient = blockchain.NewClient(n.nullBlockchain)
 		n.blockchainClient.Set(n.nullBlockchain)
 

--- a/core/blockchain/server.go
+++ b/core/blockchain/server.go
@@ -12,6 +12,8 @@
 
 package blockchain
 
+import "code.vegaprotocol.io/vega/logging"
+
 type ChainServerImpl interface {
 	ReloadConf(cfg Config)
 	Stop() error
@@ -21,12 +23,14 @@ type ChainServerImpl interface {
 // Server abstraction for the abci server.
 type Server struct {
 	*Config
+	log *logging.Logger
 	srv ChainServerImpl
 }
 
 // NewServer instantiate a new blockchain server.
-func NewServer(srv ChainServerImpl) *Server {
+func NewServer(log *logging.Logger, srv ChainServerImpl) *Server {
 	return &Server{
+		log: log,
 		srv: srv,
 	}
 }
@@ -37,6 +41,7 @@ func (s *Server) Start() error {
 
 // Stop gracefully shutdowns down the blockchain provider's server.
 func (s *Server) Stop() error {
+	s.log.Info("Stopping blockchain server")
 	return s.srv.Stop()
 }
 

--- a/core/protocol/protocol.go
+++ b/core/protocol/protocol.go
@@ -140,9 +140,9 @@ func (n *Protocol) Start() error {
 // Stop will stop all services of the protocol.
 func (n *Protocol) Stop() error {
 	// unregister conf listeners
+	n.log.Info("Stopping protocol services")
 	n.confWatcher.Unregister(n.confListenerIDs)
 	n.services.Stop()
-
 	return nil
 }
 

